File: ios/MailCore2Module/MailCore2Module.m
#import "MailCore2Module.h"
#import <MailCore/MailCore.h>

static NSString *MailCore2AddressString(MCOAddress *address) {
  if (address == nil) {
    return @"";
  }

  NSString *displayName = address.displayName ?: @"";
  NSString *mailbox = address.mailbox ?: @"";

  if (displayName.length > 0 && mailbox.length > 0) {
    return [NSString stringWithFormat:@"%@ <%@>", displayName, mailbox];
  }
  if (mailbox.length > 0) {
    return mailbox;
  }
  return displayName;
}

@implementation MailCore2Module

RCT_EXPORT_MODULE(MailCore2)

RCT_REMAP_METHOD(fetchUnread,
                 fetchUnread:(NSDictionary *)config
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSString *host = config[@"host"];
  NSNumber *portNumber = config[@"port"];
  NSString *username = config[@"username"];
  NSString *password = config[@"password"];
  NSString *security = config[@"security"];

  if (host.length == 0 || username.length == 0 || password.length == 0 || portNumber == nil) {
    reject(@"mailcore_invalid_config", @"Missing required IMAP configuration.", nil);
    return;
  }

  MCOIMAPSession *session = [[MCOIMAPSession alloc] init];
  session.hostname = host;
  session.port = portNumber.unsignedIntValue;
  session.username = username;
  session.password = password;
  session.connectionType = MCOConnectionTypeClear;

  if ([security isEqualToString:@"ssl"]) {
    session.connectionType = MCOConnectionTypeTLS;
  } else if ([security isEqualToString:@"starttls"]) {
    session.connectionType = MCOConnectionTypeStartTLS;
  }

  MCOIMAPSearchExpression *expr = [MCOIMAPSearchExpression searchUnread];
  MCOIMAPSearchOperation *searchOp = [session searchExpressionOperationWithFolder:@"INBOX" expression:expr];
  [searchOp start:^(NSError *error, MCOIndexSet *searchResult) {
    if (error != nil) {
      reject(@"mailcore_search_failed", error.localizedDescription ?: @"Search failed.", error);
      return;
    }

    if (searchResult == nil || [searchResult count] == 0) {
      resolve(@[]);
      return;
    }

    MCOIMAPFetchMessagesOperation *fetchOp = [session fetchMessagesOperationWithFolder:@"INBOX"
                                                                           requestKind:MCOIMAPMessagesRequestKindHeaders
                                                                                  uids:searchResult];
    [fetchOp start:^(NSError *fetchError, NSArray *messages, MCOIndexSet *vanishedMessages) {
      if (fetchError != nil) {
        reject(@"mailcore_fetch_failed", fetchError.localizedDescription ?: @"Fetch failed.", fetchError);
        return;
      }

      NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
      formatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
      formatter.timeZone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
      formatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss'Z'";

      NSMutableArray *results = [NSMutableArray arrayWithCapacity:messages.count];
      for (MCOIMAPMessage *message in messages) {
        MCOMessageHeader *header = message.header;
        NSString *subject = header.subject ?: @"";
        NSString *from = MailCore2AddressString(header.from);
        NSString *dateString = @"";
        if (header.date != nil) {
          dateString = [formatter stringFromDate:header.date];
        }

        [results addObject:@{
          @"id": [NSString stringWithFormat:@"%llu", message.uid],
          @"subject": subject,
          @"from": from,
          @"date": dateString,
        }];
      }

      resolve(results);
    }];
  }];
}

@end

File: android/app/src/main/java/com/inbox1/mailcore2/MailCore2Module.kt
package com.inbox1.mailcore2

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.libmailcore.Address
import com.libmailcore.ConnectionType
import com.libmailcore.IMAPFetchMessagesOperation
import com.libmailcore.IMAPMessage
import com.libmailcore.IMAPMessagesRequestKind
import com.libmailcore.IMAPSearchExpression
import com.libmailcore.IMAPSearchOperation
import com.libmailcore.IMAPSession
import com.libmailcore.MailException
import com.libmailcore.OperationCallback
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

class MailCore2Module(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
  override fun getName(): String = "MailCore2"

  @ReactMethod
  fun fetchUnread(config: ReadableMap, promise: Promise) {
    val host = config.getString("host") ?: ""
    val username = config.getString("username") ?: ""
    val password = config.getString("password") ?: ""
    val port = if (config.hasKey("port")) config.getInt("port") else -1
    val security = (config.getString("security") ?: "ssl").lowercase(Locale.US)

    if (host.isBlank() || username.isBlank() || password.isBlank() || port <= 0) {
      promise.reject("mailcore_invalid_config", "Missing required IMAP configuration.")
      return
    }

    val session = IMAPSession()
    session.setHostname(host)
    session.setPort(port)
    session.setUsername(username)
    session.setPassword(password)
    session.setConnectionType(
      when (security) {
        "ssl" -> ConnectionType.ConnectionTypeTLS
        "starttls" -> ConnectionType.ConnectionTypeStartTLS
        else -> ConnectionType.ConnectionTypeClear
      },
    )

    val searchOp: IMAPSearchOperation = session.searchOperation("INBOX", IMAPSearchExpression.searchUnread())
    searchOp.start(object : OperationCallback {
      override fun succeeded() {
        val uids = searchOp.uids()
        if (uids == null || uids.count() == 0) {
          promise.resolve(Arguments.createArray())
          return
        }

        val fetchOp: IMAPFetchMessagesOperation = session.fetchMessagesByUIDOperation(
          "INBOX",
          IMAPMessagesRequestKind.IMAPMessagesRequestKindHeaders,
          uids,
        )

        fetchOp.start(object : OperationCallback {
          override fun succeeded() {
            val messages = fetchOp.messages() ?: emptyList<IMAPMessage>()
            val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
            formatter.timeZone = TimeZone.getTimeZone("UTC")

            val results = Arguments.createArray()
            for (message in messages) {
              val header = message.header()
              val from = formatAddress(header?.from())
              val subject = header?.subject() ?: ""
              val date = header?.date()?.let { formatter.format(it) } ?: ""

              val entry = Arguments.createMap()
              entry.putString("id", message.uid().toString())
              entry.putString("from", from)
              entry.putString("subject", subject)
              entry.putString("date", date)
              results.pushMap(entry)
            }

            promise.resolve(results)
          }

          override fun failed(exception: MailException) {
            promise.reject("mailcore_fetch_failed", exception.message, exception)
          }
        })
      }

      override fun failed(exception: MailException) {
        promise.reject("mailcore_search_failed", exception.message, exception)
      }
    })
  }

  private fun formatAddress(address: Address?): String {
    if (address == null) {
      return ""
    }

    val displayName = address.displayName() ?: ""
    val mailbox = address.mailbox() ?: ""

    return when {
      displayName.isNotBlank() && mailbox.isNotBlank() -> "$displayName <$mailbox>"
      mailbox.isNotBlank() -> mailbox
      else -> displayName
    }
  }
}
